Sources/SSHDeployer/DefaultSSHClient.swift

```diff
@@ -7,18 +7,27 @@ public final class DefaultSSHClient: SSHClient {
     private let host: String
     private let port: Int32
     private let username: String
-    private let password: String
+    private let password: String?
+    private let key: String?
     
-    public init(host: String, port: Int32, username: String, password: String) throws {
+    public init(host: String, port: Int32, username: String, password: String?, key: String?) throws {
         self.host = host
         self.port = port
         self.username = username
         self.password = password
+        self.key = key
         self.ssh = try SSH(host: host, port: port)
     }
 
     public func connectAndAuthenticate() throws {
-        try ssh.authenticate(username: username, password: password)
+        if let password = password {
+            try ssh.authenticate(username: username, password: password)
+        } else if let key = key {
+            try ssh.authenticate(username: username, privateKey: key)
+        } else {
+            fatalError("Connection to \(host) has no password or privateKey")
```

Comment:
If you switch to `enum`, here you will have a exhausted `switch` with all possible combinations without `default` value, making `fatalError` not needed at all.

```swift
switch authenticationType {
    case plain(...):
        try ssh.authenticate(username: username, password: password)
    case key(...):
        try ssh.authenticate(username: username, privateKey: key)
}
```

Reply:
Also, as a side note, `fatalError` will crash the process. But it is always better to `throw` an error instead. This way it will be propagated up to the invocation site and likely reported back to the user without crashing the process. In this case, if only a single deployment destination was misconfigured, `Error` will be presented, but other deployment destinations will still be used properly and jobs will begin running.